daffodil-lib/src/main/scala/org/apache/daffodil/util/DefaultValidator.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.util
+
+import java.net.URI
+
+import com.typesafe.config.Config
+
+import scala.xml.parsing.NoBindingFactoryAdapter
+import javax.xml.XMLConstants
+import javax.xml.transform.stream.StreamSource
+import org.apache.daffodil.api.SaxValidationError
+import org.apache.daffodil.api.SaxValidationWarning
+import org.apache.daffodil.api.ValidationFailure
+import org.apache.daffodil.api.ValidationResult
+import org.apache.daffodil.api.ValidationWarning
+import org.apache.daffodil.api.Validator
+import org.apache.daffodil.api.ValidatorFactory
+import org.apache.daffodil.xml.DFDLCatalogResolver
+import org.xml.sax.ErrorHandler
+import org.xml.sax.SAXException
+import org.xml.sax.SAXParseException
+
+import scala.collection.JavaConverters._
+
+/**
+ * Use this for extra validation passes in the TDML Runner
+ * to do a validation pass on the TDML expected Infoset w.r.t. the model and to
+ * do a validation pass on the actual result w.r.t. the model as an XML document.
+ */
+class DefaultValidator(schemaFileNames: Seq[String])
+  extends NoBindingFactoryAdapter // todo;; do we need this adapter?
+  with Validator
+  with ErrorHandler {
+
+  private var errors = List.empty[ValidationFailure]
+  private var warnings = List.empty[ValidationWarning]
+
+  def validateXML(document: java.io.InputStream): ValidationResult = {
+    val schemaSources: Seq[javax.xml.transform.Source] = schemaFileNames.map { fn =>
+      val uri = new URI(fn)
+      val is = uri.toURL.openStream()
+      val stream = new StreamSource(is)
+      stream.setSystemId(uri.toString) // must set this so that relative URIs will be created for import/include files.
+      stream
+    }
+
+    val factory = new org.apache.xerces.jaxp.validation.XMLSchemaFactory()
+    factory.setErrorHandler(this)
+    val resolver = DFDLCatalogResolver.get
+    factory.setResourceResolver(resolver)
+    val schema = factory.newSchema(schemaSources.toArray)
+    val validator = initializeValidator(schema.newValidator(), this, resolver)
+
```

Comment:
Everything from here to the top of the method should happen at factory time, not at validateXML time. 

This is the change that motivates the whole design of Factory vs. Validator.

The exception here is line 68.

We may or may not need to construct a ThreadLocal that has its own schema.newValidator() per thread (for this DataProcessor object), depending on whether Xerces validation is a stateful object (I hope not), or returns everything via the result object (I hope so.)

Reply:
How about in the validator but handled at init time.  Simplifies some things this way.

Reply:
Works for me. 

Do you know if Xerces is in fact thread safe to call validate() on?

I found this:   

> You can reuse the same validator and the same schema multiple times in series. However, only the schema is     thread safe. Validators and schema factories are not. If you validate in multiple threads simultaneously, make sure each one has its own Validator and SchemaFactory objects.

On this page: https://www.ibm.com/developerworks/xml/library/x-javaxmlvalidapi/

It suggests we can't hoist out to the factory without using a thread-local for the Validator object itself.

Reply:
Everything I have read recently says it is not.  Which makes sense, the library forces clients to synchronize across threads if needed.  Which will be true of Validator instances also.

Reply:
So then my suggestion is this:

Change the validator member to:
```
def validator = validatorTL.get

val validatorTL = new ThreadLocal[XercesValidator]{
   def getInstance() = .....code to actually create one....
}
```

Reply:
hows about this

```
  private val validator = new ThreadLocal[XercesValidator] {
    override def initialValue(): XercesValidator =
      initializeValidator(schema.newValidator(), DefaultValidator.this, resolver)
  }
```

unless you had a specific reason for the def

Reply:
The reason  for the def validator, and the val (which should be private val, btw) is just to hide the fact that this is a threadlocal entirely. That's the implementation's business, not the user of the validator. The user of the validator shouldn't have to know that it's a threadlocal and so .get is needed, etc. 

If you put the types on the rationale is clear:
```
def validator: XercesValidator = validatorTL.get

private val validatorTL = new ThreadLocal[XercesValidator] { ... }
```
Unless the caller of the validator.validateXML method looks in the implementation, they don't know anything about this thread-local stuff.

Reply:
The reason  for the def validator, and the val (which should be private val, btw) is just to hide the fact that this is a threadlocal entirely. That's the implementation's business, not the user of the validator. The user of the validator shouldn't have to know that it's a threadlocal and so .get is needed, etc. 

If you put the types on the rationale is clear:
```
def validator: XercesValidator = validatorTL.get

private val validatorTL = new ThreadLocal[XercesValidator] { ... }
```
Unless the caller of the validator.validateXML method looks in the implementation, they don't know anything about this thread-local stuff.

Reply:
Are you requesting changes to the current code?  Im not following why we want an an accessor thats not exposed by the trait, to access a private var that users dont need to know about.

Reply:
Yeah. You are right. I was somehow thinking this validator got exposed by the object, but it's only used locally inside the validateXML method. Nevermind.

Reply:
:+1: